
<!-- README.md is generated from README.Rmd. Please edit that file -->

# toolkit4pySCA

<!-- badges: start -->

[![R-CMD-check](https://github.com/currocam/toolkit4pySCA/workflows/R-CMD-check/badge.svg)](https://github.com/currocam/toolkit4pySCA/actions)
<!-- badges: end -->

The goal of toolkit4pySCA is to provide a bunch of useful functions in
order to obtain a sequence alignment using HMMER and associated
taxonomic information.Then, it allows to read the file generated by
pySCA and to analyze the results easily with R and tidyverse.

## Installation

You can install the development version of toolkit4pySCA from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("currocam/toolkit4pySCA")
```

## Example

This is a basic example showing how to carry out a typical workflow.

### phmmer

First, we performed a search for sequences homologous to the sequence of
interest using phmmer, indicating the sequence of interest and the
database. This can be done using `quick_AA_search_using_phmmer`,
`download_xml_from_phmer()` or downloading the xml file from the web
server. Actually, programmatic access to the HMMER only seems to make
sense if you need to perform a lot of queries and want to automate the
process.

``` r
library(toolkit4pySCA)
progressr::handlers(global = TRUE)

db <- c("swissprot", "pdb")
# >2abl_A mol:protein length:163  ABL TYROSINE KINASE
example.seq <- AAString(
  paste0("MGPSENDPNLFVALYDFVASGDNTLSITKGEKLRVLGYNHNGEWCEAQTKNGQGWVPSNYITPVNSLE",
         "KHSWYHGPVSRNAAEYLLSSGINGSFLVRESESSPGQRSISLRYEGRVYHYRINTASDGKLYVSSESRFNTLAELV",
         "HHHSTVADGLITTLHYPAP")
df.phmmer <- quick_AA_search_using_phmmer(example.seq, db)
```

``` r
library(Biostrings)
library(xml2)
fasta.files <- readAAStringSet(fullseq.fasta.files)
df.phmmer.info <- xml.files %>%
  purrr::map_dfr(
    ~read_xml(.x) %>%
      extract_tidy_df_from_hmmer,
    .id = "db")
```

``` r
library(dplyr)
hits <- extract_tidy_hits_from_xml(xml.document)
stats <- extract_tidy_stats_from_xml(xml.document)
hits%>%
  select(pvalue, evalue, score) %>%
  summary()
```

    library(dplyr)
    hits <- df.phmmer.info%>%
      group_by(alisqacc) 
    hits %>%
      select(alisqacc, pvalue, evalue, score) %>%
      summary()

<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
alisqacc
</th>
<th style="text-align:left;">
pvalue
</th>
<th style="text-align:left;">
evalue
</th>
<th style="text-align:left;">
score
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Length:1028
</td>
<td style="text-align:left;">
Min. :-265.00
</td>
<td style="text-align:left;">
Min. :0.00e+00
</td>
<td style="text-align:left;">
Min. : 20.90
</td>
</tr>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Class :character
</td>
<td style="text-align:left;">
1st Qu.: -54.51
</td>
<td style="text-align:left;">
1st Qu.:0.00e+00
</td>
<td style="text-align:left;">
1st Qu.: 31.10
</td>
</tr>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Mode :character
</td>
<td style="text-align:left;">
Median : -36.57
</td>
<td style="text-align:left;">
Median :0.00e+00
</td>
<td style="text-align:left;">
Median : 47.20
</td>
</tr>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Mean : -44.22
</td>
<td style="text-align:left;">
Mean :4.12e-04
</td>
<td style="text-align:left;">
Mean : 57.99
</td>
</tr>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
3rd Qu.: -25.20
</td>
<td style="text-align:left;">
3rd Qu.:6.40e-06
</td>
<td style="text-align:left;">
3rd Qu.: 72.50
</td>
</tr>
<tr>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Max. : -17.97
</td>
<td style="text-align:left;">
Max. :8.90e-03
</td>
<td style="text-align:left;">
Max. :369.80
</td>
</tr>
</tbody>
</table>

``` r
library(forcats)
hits %>%
  mutate(
    ph = fct_lump(ph, 2),
  ) %>%
  pull(ph)%>%
  fct_count(prop = TRUE, sort = TRUE) %>%
  kbl() %>%
  kable_styling(position = "center")
```

<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
f
</th>
<th style="text-align:right;">
n
</th>
<th style="text-align:right;">
p
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Chordata
</td>
<td style="text-align:right;">
894
</td>
<td style="text-align:right;">
0.8696498
</td>
</tr>
<tr>
<td style="text-align:left;">
Ascomycota
</td>
<td style="text-align:right;">
36
</td>
<td style="text-align:right;">
0.0350195
</td>
</tr>
<tr>
<td style="text-align:left;">
Nematoda
</td>
<td style="text-align:right;">
27
</td>
<td style="text-align:right;">
0.0262646
</td>
</tr>
<tr>
<td style="text-align:left;">
Artverviricota
</td>
<td style="text-align:right;">
26
</td>
<td style="text-align:right;">
0.0252918
</td>
</tr>
<tr>
<td style="text-align:left;">
Arthropoda
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
0.0194553
</td>
</tr>
<tr>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0.0068093
</td>
</tr>
<tr>
<td style="text-align:left;">
Evosea
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0048638
</td>
</tr>
<tr>
<td style="text-align:left;">
Cnidaria
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0029183
</td>
</tr>
<tr>
<td style="text-align:left;">
Basidiomycota
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0029183
</td>
</tr>
<tr>
<td style="text-align:left;">
Porifera
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0029183
</td>
</tr>
<tr>
<td style="text-align:left;">
Discosea
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0.0019455
</td>
</tr>
<tr>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0.0019455
</td>
</tr>
</tbody>
</table>

``` r
library(ggplot2)
library(tidyr)
library(dplyr)

df.phmmer.info%>%
  filter(evalue < 0.001) %>%
  mutate(
  desc = fct_lump_n(desc, 5),
  ph = fct_lump(ph, 2),
  )%>%
  group_by(ph, desc) %>%
  nest() %>%
  ggplot(aes(y=desc, fill = ph)) + 
    geom_histogram(stat = "count")
```

<img src="man/figures/README-hist_ph_desc_hmmer-1.png" width="100%" />

``` r
tax.df <- hits %>%
    pull(taxid) %>%
    download_tax_from_ncbi() %>%
    mutate(taxid = as.numeric(taxid))
#> Registered S3 method overwritten by 'hoardr':
#>   method           from
#>   print.cache_info httr
#> 
#> Attaching package: 'purrr'
#> The following object is masked from 'package:XVector':
#> 
#>     compact
#> The following object is masked from 'package:IRanges':
#> 
#>     reduce

hits %>%
  left_join(tax.df, by = c("taxid" = "taxid"))
#> # A tibble: 1,028 × 88
#> # Groups:   alisqacc [848]
#>    db    aliId aliIdCount  aliL  aliM  aliN aliSim aliSimCount aliaseq alicsline
#>    <chr> <dbl>      <dbl> <dbl> <dbl> <dbl>  <dbl>       <dbl> <chr>   <lgl>    
#>  1 1     1            163   163   163   163  1             163 MGPSEN… NA       
#>  2 1     1            162   495   163   162  1             162 GPSEND… NA       
#>  3 1     1            162   495   163   162  1             162 GPSEND… NA       
#>  4 1     1            162   495   163   162  1             162 GPSEND… NA       
#>  5 1     1            162   537   163   162  1             162 GPSEND… NA       
#>  6 1     0.994        161   255   163   162  1             162 GPSEND… NA       
#>  7 1     1            107   123   163   107  1             107 PSNYIT… NA       
#>  8 1     1            107   121   163   107  1             107 PSNYIT… NA       
#>  9 1     1             99   109   163    99  1              99 NSLEKH… NA       
#> 10 1     0.990        100   408   163   101  0.990         100 SVNSLE… NA       
#> # … with 1,018 more rows, and 78 more variables: alihmmacc <lgl>,
#> #   alihmmdesc <lgl>, alihmmfrom <dbl>, alihmmname <chr>, alihmmto <dbl>,
#> #   alimline <chr>, alimmline <lgl>, alimodel <chr>, aliniseqs <dbl>,
#> #   alintseq <lgl>, alippline <chr>, alirfline <lgl>, alisindex <dbl>,
#> #   alisqacc <chr>, alisqdesc <chr>, alisqfrom <dbl>, alisqname <chr>,
#> #   alisqto <dbl>, bias.x <dbl>, bitscore <dbl>, cevalue <dbl>, iali <dbl>,
#> #   ienv <dbl>, ievalue <dbl>, is_included <dbl>, is_reported <dbl>, …
```

``` r
pdb.id <- "2ABL"
pdb.url <- paste0("https://files.rcsb.org/download/",
                  pdb.id, ".pdb")
curl::curl_download(pdb.url, paste0(path, pdb.id, ".pdb"))
```

``` bash
scaProcessMSA -a 2abl_A_significant.ClustalW.aln -s 2ABL 

Loaded alignment of 602 sequences, 3039 positions.
Checking alignment for non-standard amino acids
Aligment size after removing sequences with non-standard amino acids: 600
Trimming alignment for highly gapped positions (80% or more).
Alignment size post-trimming: 625 positions
Looking for PDBs in .
Finding reference sequence using global MSAsearch...
Trying MSASearch with ggsearch
reference sequence index is: 296
2abl_A
----------------------------------------------------------------M-GPSENDPNVALYDFVASGDNTLSITKGEKHNGEWCEAQTKN-QGWVPSNYITPVNSLEKHSWYHGPVSRNAAEYLLSSGINGSFLVRESESSPGQRSISLRYEGR--------VYHYRINTASDGKLYVSSESRFNTLAELVHHHSTVAD--GLITTLHYPAP------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
len refseq 163, len refpos 163, len algseq 625, len pairalg 635, len gloalg 625
Conducting sequence and position filtering: alignment size is 600 seqs, 625 pos
ATS and distmat size - ATS: 625, distmat: 625 x 625
Keeping 60 sequences of 600 sequences (after filtering for gaps)
Keeping 46 sequences of 60 sequences (after filtering for seq similarity)
After filtering: alignment size is 46 seqs, 11 effective seqs, 504 pos
Final alignment parameters:
Number of sequences: M = 46
Number of effective sequences: M' = 11
Number of alignment positions: L = 504
Number of positions in the ats: 504
Number of structure positions mapped: 152
Size of the distance matrix: 504 x 504
Opening database file 

scaCore -i 2abl_A_significant.ClustalW.db

omputing the sequence projections.
basicICA failed to converge: 36.454102510130255
basicICA failed to converge: 50.01948709627572
basicICA failed to converge: 9.683913030829293
Computing sequence similarity matrix.
Computing the SCA conservation and correlation values.
Computing matrix randomizations...
Randomizations complete, 10 trials, time: 1.4 minutes
Calculations complete, writing to database file

scaSectorID -i 2abl_A_significant.ClustalW.db 

Selected kpos=5 significant eigenmodes.
Calculations complete, writing to database file
```

``` r
pysca <- import_pySCA_module_using_reticulate( 
  condaenv = "r-pySCA",force_installing = TRUE
  )
#reticulate::py_module_available(c("Bio"))
#db <- reticulate::py_load_object("inst/extdata/2abl_A_significant.ClustalW.db", pickle = "pickle")
```
